text/0000-cargo-workspace.md

```diff
@@ -0,0 +1,362 @@
+`[replace]` to exist in the workspace root. This Cargo.toml will affect lock
+file generation, but no other workspace members will be allowed to have a
+`[replace]` directive (with an informative error message being produced).
+
+Finally, the `paths` overrides will be applied as usual, and they'll continue to
+be applied relative to whatever crate is being compiled (not the workspace
+root). These are intended for much more local testing, so no restriction of
+"must be in the root" should be necessary.
+
+### Future Extensions
+
+Once Cargo understands a workspace of crates, we could easily extend various
+subcommands with a `--all` flag to perform tasks such as:
+
+* Test all crates within a workspace (run all unit tests, doc tests, etc)
+* Build all binaries for a set of crates within a workspace
+* Publish all crates in a workspace if necessary to crates.io
+
+This support isn't proposed to be added in this RFC specifically, but simply to
+show that workspaces can be used to solve other existing issues in Cargo.
+
+# Drawbacks
+
+* This change is not backwards compatible with older versions of Cargo.lock. For
+  example if a newer cargo were used to develop a repository which otherwise is
+  developed with older versions of Cargo, the `Cargo.lock` files generated would
+  be incompatible. If all maintainers agree on versions of Cargo, however, this
+  is not a problem.
+
+* As proposed there is no method to disable implicit actions taken by Cargo.
+  It's unclear what the use case for this is, but it could in theory arise.
+
+* No crate will implicitly benefit from workspaces after this is implemented.
+  Existing crates must opt-in with a `[workspace]` key somewhere at least.
+
+# Alternatives
+
+* The `workspace.members` key could support globs to define a number of
+  directories at once. For example one could imagine:
+
+  ```toml
+  [workspace]
+  members = ["crates/*"]
+  ```
+
+  as an ergonomic method of slurping up all sub-folders in the `crates` folder
+  as crates.
+
+* Cargo could attempt to perform more inference of workspace members by simply
+  walking the entire directory tree starting at `Cargo.toml`. All children found
+  could implicitly be members of the workspace. Walking entire trees,
+  unfortunately, isn't always efficient to do and it would be unfortunate to
+  have to unconditionally do this.
+
+* Cargo could support "virtual packages" where a `Cargo.toml` is placed at the
+  root of a repository but only to serve as a global project configuration. No
+  crate would actually be described by a virtual package, but it would play into
+  the workspace heuristics described here. This feature could alleviate the "too
+  much extra configuration" drawback described above, but it's unclear whether
+  it's needed at this point.
+
+* Implicit members are currently only path dependencies and a "Cargo.toml next
+  to VCS" traveling upwards. Instead all Cargo.toml members found traveling
```

Comment:
"upmost Cargo.toml between VCS root and current directory, inclusive" I believe is the current rule?